overlord/snapstate/autorefresh_gating_test.go

```diff
@@ -211,6 +223,392 @@ slots:
+		},
+	})
+}
+
+func (s *autorefreshGatingSuite) TestHoldRefreshHelperCloseToMaxPostponement(c *C) {
+	st := s.state
+	st.Lock()
+	defer st.Unlock()
+
+	lastRefreshedStr := "2021-01-01T10:00:00Z"
+	lastRefreshed, err := time.Parse(time.RFC3339, lastRefreshedStr)
+	c.Assert(err, IsNil)
+	// we are 1 day before maxPostponent
+	now := lastRefreshed.Add(89 * time.Hour * 24)
+
+	restore := snapstate.MockTimeNow(func() time.Time { return now })
+	defer restore()
+
+	mockInstalledSnap(c, st, snapAyaml, false)
+	mockInstalledSnap(c, st, snapByaml, false)
+	mockLastRefreshed(c, st, lastRefreshedStr, "snap-a")
+
+	// request default hold time
+	var hold time.Duration
+	c.Assert(snapstate.HoldRefresh(st, "snap-b", hold, "snap-a"), IsNil)
+
+	var gating map[string]map[string]*snapstate.HoldState
+	c.Assert(st.Get("snaps-hold", &gating), IsNil)
+	c.Assert(gating, HasLen, 1)
+	c.Check(gating["snap-a"]["snap-b"].HoldUntil.String(), DeepEquals, lastRefreshed.Add(90*time.Hour*24).String())
+}
+
+func (s *autorefreshGatingSuite) TestHoldRefreshExplicitHoldTime(c *C) {
+	st := s.state
+	st.Lock()
+	defer st.Unlock()
+
+	now := "2021-05-10T10:00:00Z"
+	restore := snapstate.MockTimeNow(func() time.Time {
+		t, err := time.Parse(time.RFC3339, now)
+		c.Assert(err, IsNil)
+		return t
+	})
+	defer restore()
+
+	mockInstalledSnap(c, st, snapAyaml, false)
+	mockInstalledSnap(c, st, snapByaml, false)
+
+	hold := time.Hour * 24 * 3
+	// holding self for 3 days
+	c.Assert(snapstate.HoldRefresh(st, "snap-a", hold, "snap-a"), IsNil)
+
+	// snap-b holds snap-a for 1 day
+	hold = time.Hour * 24
+	c.Assert(snapstate.HoldRefresh(st, "snap-b", hold, "snap-a"), IsNil)
+
+	var gating map[string]map[string]*snapstate.HoldState
+	c.Assert(st.Get("snaps-hold", &gating), IsNil)
+	c.Check(gating, DeepEquals, map[string]map[string]*snapstate.HoldState{
+		"snap-a": {
+			"snap-a": snapstate.MockHoldState(now, "2021-05-13T10:00:00Z"),
+			"snap-b": snapstate.MockHoldState(now, "2021-05-11T10:00:00Z"),
+		},
+	})
+}
+
+func (s *autorefreshGatingSuite) TestHoldRefreshHelperErrors(c *C) {
+	st := s.state
+	st.Lock()
+	defer st.Unlock()
+
+	now := "2021-05-10T10:00:00Z"
+	restore := snapstate.MockTimeNow(func() time.Time {
+		t, err := time.Parse(time.RFC3339, now)
+		c.Assert(err, IsNil)
+		return t
+	})
+	defer restore()
+
+	mockInstalledSnap(c, st, snapAyaml, false)
+	mockInstalledSnap(c, st, snapByaml, false)
+	// snap-b was refreshed a few days ago
+	mockLastRefreshed(c, st, "2021-05-01T10:00:00Z", "snap-b")
+
+	// holding itself
+	hold := time.Hour * 24 * 96
+	c.Assert(snapstate.HoldRefresh(st, "snap-a", hold, "snap-a"), ErrorMatches, `cannot hold some snaps:\n - requested holding duration 2304h0m0s exceeds maximum holding for snap "snap-a"`)
+
+	// holding other snap
+	hold = time.Hour * 49
+	err := snapstate.HoldRefresh(st, "snap-a", hold, "snap-b")
+	c.Check(err, ErrorMatches, `cannot hold some snaps:\n - requested holding duration 49h0m0s exceeds maximum holding for snap "snap-b"`)
```

Comment:
I'm not sure this error message is the most clear, in this instance we had `snap-b` trying to hold back refresh of `snap-a`, but it failed because it tried to postpone it for 49 hours, which is more than the max, I would think a more clear error message would be something like:

```suggestion
	c.Check(err, ErrorMatches, `cannot hold some snaps:\n - requested holding duration for "snap-a" of 49h0m0s by "snap-b" exceeds maximum holding time`)
```

or you could get rid of the `time` at the end of the error, but I think this makes it more clear the direction of holding in the error message